lib/paperclip/storage/s3.rb

```diff
@@ -0,0 +1,34 @@
+# This module adds the ability for Paperclip to work with AWS STS.
+# Once we migrate from Paperclip to ActiveStorage, this file should be removed.
+module Paperclip
+  module Storage
+    module S3
+      def s3_interface
+        @s3_interface ||= begin
+          config = { region: s3_region }
+
+          if using_http_proxy?
+
+            proxy_opts = { :host => http_proxy_host }
+            proxy_opts[:port] = http_proxy_port if http_proxy_port
+            if http_proxy_user
+              userinfo = http_proxy_user.to_s
+              userinfo += ":#{http_proxy_password}" if http_proxy_password
+              proxy_opts[:userinfo] = userinfo
+            end
+            config[:proxy_uri] = URI::HTTP.build(proxy_opts)
+          end
+
+          config[:use_accelerate_endpoint] = use_accelerate_endpoint?
+
+          # Add :session_token as a valid argument
+          [:access_key_id, :secret_access_key, :credential_provider, :credentials, :session_token].each do |opt|
```

Comment:
**If I remember correctly**, `session_token` is the temporary thing that needs to be updated so we don't like to use it directly. I think there was a credentials class that will read token regularly from a file so we can just pass the `credentials` object and let it figure out when and how to read the token.

In this way no monkey patching would be needed:
https://github.com/thoughtbot/paperclip/blob/main/lib/paperclip/storage/s3.rb#L251

Reply:
And I think you're right... Paperclip [instantiates Aws::S3::Resource](https://github.com/thoughtbot/paperclip/blob/c769382c9b7078f3d1620b50ec2a70e91ba62ec4/lib/paperclip/storage/s3.rb#L261) which, in turn, [instantiates Aws::S3::Client](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/S3/Client.html#initialize-instance_method). Just passing a hash inside of credentials with `:access_key_id`, `:secret_access_key` and `:session_token`, could be enough.
Let me check that out 👀

Reply:
Actually it is in my JIRA comment. I believe you need to create a `AssumeRoleWebIdentityCredential` and pass it as the `credentials` parameter.

Reply:
Following up on the tests:

- Passing `credentials` down to `Aws::S3::Resource` in the hopes to avoid the monkey patch
    ```ruby
    sts = Aws::STS::Client.new(CMS::S3.credentials.merge(region: CMS::S3.region))
    session = sts.get_session_token(duration_seconds: 15*60)
    resource = Aws::S3::Resource.new(credentials: session, region: CMS::S3.region)
    
    resource.client.get_bucket_tagging(bucket: 'edge-flexo.3scale.net.3scale.net') #=> Works
    ```

I was checking [Aws::STS::Client#assume_role_with_web_identity](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/STS/Client.html#assume_role_with_web_identity-instance_method) and, since it also returns the credentials, we should be OK to use it in the same fashion as the example above. Now, for us to do that, we'll need more params that I'm not really sure we have or where to get from. Namely `role_arn`, `role_session_name` and `web_identity_token`:
```ruby
resp = client.assume_role_with_web_identity({
  role_arn: "arnType", # required
  role_session_name: "roleSessionNameType", # required
  web_identity_token: "clientTokenType", # required
  provider_id: "urlType",
  policy_arns: [
    {
      arn: "arnType",
    },
  ],
  policy: "sessionPolicyDocumentType",
  duration_seconds: 1,
})
```
Any ideas on those?

Reply:
I don't know about the other parameters. Hopefully whoever requested that, will know. The point of `AssumeRoleWebIdentityCredential` is that it should update automatically when token expires. Otherwise we will have to implement that somehow.

Reply:
> The point of AssumeRoleWebIdentityCredential is that it should update automatically when token expires.

Yes. But from what I got from the documentation, this should also be true for many of the other methods (APIs) available in `Aws::STS::Client`, such as `assume_role` and `get_session_token`. What's still unclear to me is how the "mounted file that contains signed token(which is automatically rotated every few minutes)" (described in the ticket) would play a part in here 🤔

Reply:
I haven't looked now, from my memory only `AssumeRoleWebIdentityCredential` reads token from a file the way feature is described. All other options need  other credentials or don't update automatically when token changed.

Reply:
And you're right yet again!
I just double checked the [AssumeRoleWebIdentityCredentials](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/AssumeRoleWebIdentityCredentials.html) **class** and it accepts a `web_identity_token_file` param in its initializer that sounds closer to what the ticket says. I'll keep this one close to mind.
Thank you!